app/master/build_scheduler.py

```diff
@@ -0,0 +1,122 @@
+from queue import Empty
+from threading import Lock
+
+from app.master.slave import SlaveMarkedForShutdownError
+from app.util import analytics
+from app.util.log import get_logger
+
+# pylint: disable=protected-access
+# Disable protected-access for whole file; will be fixing this (by
+# moving Build's subjob queues into this class) in a separate change.
+
+
+class BuildScheduler(object):
+    """
+    This class handles the logic of taking a Build and distributing its subjobs to
+    slaves. There is a one-to-one relationship between instances of this class and
+    instances of Build.
+
+    The data flow between this class and Build is unidirectional; the build instance
+    itself shouldn't know anything about its scheduler or slaves or where subjobs
+    are executing. All that goes here.
+
+    Ususally this class is instantiated and managed by a BuildSchedulerPool.
+    """
+    def __init__(self, build):
+        """
+        :type build: Build
+        """
+        self._logger = get_logger(__name__)
+        self._build = build
+
+        job_config = build.project_type.job_config()
+        self._max_executors = job_config.max_executors
+        self._max_executors_per_slave = job_config.max_executors_per_slave
+
+        self._slaves_allocated = []
+        self._num_executors_allocated = 0
+        self._num_executors_in_use = 0
+        self._subjob_assignment_lock = Lock()  # prevents subjobs from being skipped
+
+    @property
+    def build_id(self):
+        """ :rtype: int """
+        return self._build.build_id()
+
+    def needs_more_slaves(self):
+        """
+        Determine whether or not this build should have more slaves allocated to it.
+
+        :rtype: bool
+        """
+        return self._num_executors_allocated < self._max_executors and not self._build._unstarted_subjobs.empty()
```

Comment:
The `self._build._unstarted_subjobs` reference is a temporary violation of the build's public interface. I plan to fix this in a future change, probably by moving the subjob queues into this class.

Some of the state computations in the `Build` class still depend on these queues so it's not as easy to remove them into this class before the state machine refactor.